org.elasticsearch.index.reindex.ReindexClientYamlTestSuiteIT.test {yaml=reindex/90_remote/Reindex from remote that you can't connect to} fails on Windows

Build URL: http://build-us-00.elastic.co/job/es_core_master_windows-2012-r2/4418/consoleFull

This does not reproduce on my machine (maybe because this only reproduces on windows) and because its a YAML test unfortunately does not show any server stack trace. It seems to have been failing since 9th December.

@nik9000 would you be able to take a look?

Reproduce command:
```
gradle :modules:reindex:integTest -Dtests.seed=32E1F100D3E3755 -Dtests.class=org.elasticsearch.index.reindex.ReindexClientYamlTestSuiteIT -Dtests.method="test {yaml=reindex/90_remote/Reindex from remote that you can't connect to}" -Dtests.es.logger.level=DEBUG -Dtests.assertion.disabled=org.elasticsearch -Dtests.security.manager=true -Dtests.nightly=false -Dtests.heap.size=512m -Dtests.jvm.argline="-server -XX:+UseConcMarkSweepGC -XX:+UseCompressedOops -XX:+AggressiveOpts" -Dtests.locale=nl-NL -Dtests.timezone=America/Hermosillo
```

Error:
```
1> [2016-12-15T04:57:13,845][INFO ][o.e.i.r.ReindexClientYamlTestSuiteIT] Stash dump on failure [{
  1>   "stash" : {
  1>     "body" : {
  1>       "error" : {
  1>         "root_cause" : [
  1>           {
  1>             "type" : "null_pointer_exception",
  1>             "reason" : "charset"
  1>           }
  1>         ],
  1>         "type" : "null_pointer_exception",
  1>         "reason" : "charset"
  1>       },
  1>       "status" : 500
  1>     }
  1>   }
  1> }]

  2> REPRODUCE WITH: gradle :modules:reindex:integTest -Dtests.seed=32E1F100D3E3755 -Dtests.class=org.elasticsearch.index.reindex.ReindexClientYamlTestSuiteIT -Dtests.method="test {yaml=reindex/90_remote/Reindex from remote that you can't connect to}" -Dtests.es.logger.level=DEBUG -Dtests.assertion.disabled=org.elasticsearch -Dtests.security.manager=true -Dtests.nightly=false -Dtests.heap.size=512m -Dtests.jvm.argline="-server -XX:+UseConcMarkSweepGC -XX:+UseCompressedOops -XX:+AggressiveOpts" -Dtests.locale=nl-NL -Dtests.timezone=America/Hermosillo

FAILURE 1.79s | ReindexClientYamlTestSuiteIT.test {yaml=reindex/90_remote/Reindex from remote that you can't connect to} <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: Failure at [reindex/90_remote:303]: the error message was expected to match the provided regex but didn't
   > Expected: Connection refused
   >      but: was "{root_cause=[{type=null_pointer_exception, reason=charset}], type=null_pointer_exception, reason=charset}"
   > 	at __randomizedtesting.SeedInfo.seed([32E1F100D3E3755:8B7A20CAA3C25AAD]:0)
   > 	at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:374)
   > 	at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.test(ESClientYamlSuiteTestCase.java:354)
   > 	at java.lang.Thread.run(Thread.java:745)
   > Caused by: java.lang.AssertionError: the error message was expected to match the provided regex but didn't
   > Expected: Connection refused
   >      but: was "{root_cause=[{type=null_pointer_exception, reason=charset}], type=null_pointer_exception, reason=charset}"
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.test.rest.yaml.section.DoSection.execute(DoSection.java:156)
   > 	at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:370)
   > 	... 37 more
```

--- a/client/rest/src/main/java/org/elasticsearch/client/RequestLogger.java
+++ b/client/rest/src/main/java/org/elasticsearch/client/RequestLogger.java
@@ -152,7 +152,7 @@ final class RequestLogger {
             httpResponse.setEntity(entity);
             ContentType contentType = ContentType.get(entity);
             Charset charset = StandardCharsets.UTF_8;
-            if (contentType != null) {
+            if (contentType != null && contentType.getCharset() != null) {
                 charset = contentType.getCharset();
             }
             try (BufferedReader reader = new BufferedReader(new InputStreamReader(entity.getContent(), charset))) {

--- a/client/rest/src/test/java/org/elasticsearch/client/RequestLoggerTests.java
+++ b/client/rest/src/test/java/org/elasticsearch/client/RequestLoggerTests.java
@@ -44,6 +44,7 @@ import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
 
 import static org.hamcrest.CoreMatchers.equalTo;
@@ -51,7 +52,6 @@ import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertThat;
 
 public class RequestLoggerTests extends RestClientTestCase {
-
     public void testTraceRequest() throws IOException, URISyntaxException {
         HttpHost host = new HttpHost("localhost", 9200, randomBoolean() ? "http" : "https");
         String expectedEndpoint = "/index/type/_api";
@@ -69,7 +69,7 @@ public class RequestLoggerTests extends RestClientTestCase {
             expected += " -d '" + requestBody + "'";
             HttpEntityEnclosingRequest enclosingRequest = (HttpEntityEnclosingRequest) request;
             HttpEntity entity;
-            switch(randomIntBetween(0, 3)) {
+            switch(randomIntBetween(0, 4)) {
                 case 0:
                     entity = new StringEntity(requestBody, StandardCharsets.UTF_8);
                     break;
@@ -81,6 +81,10 @@ public class RequestLoggerTests extends RestClientTestCase {
                     break;
                 case 3:
                     entity = new NByteArrayEntity(requestBody.getBytes(StandardCharsets.UTF_8));
+                    break;
+                case 4:
+                    // Evil entity without a charset
+                    entity = new StringEntity(requestBody, (Charset) null);
                     break;
                 default:
                     throw new UnsupportedOperationException();
@@ -116,11 +120,20 @@ public class RequestLoggerTests extends RestClientTestCase {
             expected += "\n#   \"field\": \"value\"";
             expected += "\n# }";
             HttpEntity entity;
-            if (getRandom().nextBoolean()) {
+            switch(randomIntBetween(0, 2)) {
+            case 0:
                 entity = new StringEntity(responseBody, StandardCharsets.UTF_8);
-            } else {
+                break;
+            case 1:
                 //test a non repeatable entity
                 entity = new InputStreamEntity(new ByteArrayInputStream(responseBody.getBytes(StandardCharsets.UTF_8)));
+                break;
+            case 2:
+                // Evil entity without a charset
+                entity = new StringEntity(responseBody, (Charset) null);
+                break;
+            default:
+                throw new UnsupportedOperationException();
             }
             httpResponse.setEntity(entity);
         }

